File: 0113-path-sum-ii/0113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<List<Integer>>ans=new ArrayList<>();
        if(root==null){
            return ans;
        }
        int sum=0;
        List<Integer>res=new ArrayList<>();
        recursion(root,ans,res,sum,targetSum);
        return ans;
        
    }
    public void recursion(TreeNode root,List<List<Integer>>ans,List<Integer>res,int sum,int targetSum){
        if(root==null){
            return;
        }
        sum+=root.val;
        // if(sum>targetSum){
        //     return;
        // }
        res.add(root.val);
        recursion(root.left,ans,res,sum,targetSum);
        recursion(root.right,ans,res,sum,targetSum);
        if(root.left==null && root.right==null){
            if(sum==targetSum){
                ans.add(new ArrayList<>(res));
            }
        }
        res.remove(res.size()-1);

    }
}

File: 0113-path-sum-ii/0113-path-sum-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        if root is None:
            return []
        self.res=[]
        self.solve(root,targetSum,[])
        return self.res
    def solve(self,root,targetSum,temp):
        if root is None:
            return 
        if root.left is None and root.right is None:
            if targetSum-root.val==0:
                temp.append(root.val)
                self.res.append(temp.copy())
                temp.pop()
                return 
        temp.append(root.val)
        if root.left:
            self.solve(root.left,targetSum-root.val,temp)
        if root.right:
            self.solve(root.right,targetSum-root.val,temp)
        temp.pop()
        

        